### PR TITLE
feat(general): enrich terraform definitions context key

### PR DESCRIPTION
**Pull request by lirshindalman**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

in terraform definitions_context change the key to all TFDefinitionKey objects instead of only the file_path

## Description

*Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.*

Fixes # (issue)

### Description
*Include a description of what makes it a violation and any relevant external links.*

### Fix
*How does someone fix the issue in code and/or in runtime?*

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my feature, policy, or fix is effective and works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: checkov/terraform/runner.py [hunk 1]
```diff
@@ -61,6 +61,7 @@
 dpath.options.ALLOW_EMPTY_STRING_KEYS = True
 
 CHECK_BLOCK_TYPES = frozenset(['resource', 'data', 'provider', 'module'])
+ENABLE_DEFINITION_KEY = strtobool(os.getenv('ENABLE_DEFINITION_KEY', 'False'))
```
Comment (omryMen) [inline checkov/terraform/runner.py:64]: I'm not sure about it - but if we will do the import of this file before creating this env var it might not work

File: checkov/terraform/runner.py [hunk 2]
```diff
@@ -374,21 +375,22 @@ def run_block(self, entities,
...
-                    except ValueError as e:
-                        # TODO handle multiple modules with the same name in repo
-                        logging.warning(f'Failed to get module name for resource {entity_id}. {str(e)}')
-                        continue
-                    module_name = full_definition_path[module_name_index]
-                    caller_context = definition_context[module].get(BlockType.MODULE, {}).get(module_name)
+                module_full_path, _, module_name = get_module_from_full_path(full_file_path)
+                if module_full_path:
+                    if not module_name:
+                        full_definition_path = entity_id.split('.')
```
Comment (bo156) [inline checkov/terraform/runner.py:380]: can you split this block to a function like `_get_module_name`

**Review by bo156 — CHANGES_REQUESTED** [review]

File: checkov/common/util/parser_utils.py [hunk 3]
```diff
@@ -385,22 +388,29 @@ def get_tf_definition_key_from_module_dependency(
...
-        return file_path.tf_source_modules.path, None
+            return None, None, None
+        if ENABLE_DEFINITION_KEY:
+            modules_full_key = TFDefinitionKey(file_path=file_path.tf_source_modules.path, tf_source_modules=file_path.tf_source_modules.nested_tf_module)
+        else:
+            modules_full_key = file_path.tf_source_modules.path
+        module_name = file_path.tf_source_modules.name
+        if file_path.tf_source_modules.foreach_idx:
+            module_name = f'{module_name}[\"{file_path.tf_source_modules.foreach_idx}\"]'
+        return modules_full_key, None, module_name
```
[the remark this answers is not in the document]
Comment (lirshindalman) [reply checkov/common/util/parser_utils.py:404]: I add the module_name and tried to leave the function as it was. if its not TFDefinitionKey we return index and not None

[the remark this answers is not in the document]
Comment (ChanochShayner) [reply checkov/common/util/parser_utils.py:404]: This func should return 2 vars - we are using it in some places that expecting 2 vars.

File: checkov/common/util/parser_utils.py [hunk 4]
```diff
@@ -385,22 +388,29 @@ def get_tf_definition_key_from_module_dependency(
...
-            return None, None
-        return file_path.tf_source_modules.path, None
+            return None, None, None
+        if ENABLE_DEFINITION_KEY:
+            modules_full_key = TFDefinitionKey(file_path=file_path.tf_source_modules.path, tf_source_modules=file_path.tf_source_modules.nested_tf_module)
+        else:
+            modules_full_key = file_path.tf_source_modules.path
+        module_name = file_path.tf_source_modules.name
+        if file_path.tf_source_modules.foreach_idx:
+            module_name = f'{module_name}[\"{file_path.tf_source_modules.foreach_idx}\"]'
```
[the remark this answers is not in the document]
Comment (lirshindalman) [reply checkov/common/util/parser_utils.py:403]: until now we extract the module_name from the entity_id. It makes more sense to me to extract the name from the TFDefinitionKey if we have it

**Review by omryMen — APPROVED** [review]

File: checkov/terraform/context_parsers/base_parser.py [hunk 5]
```diff
@@ -154,7 +155,7 @@ def _compute_definition_end_line(self, start_line_num: int) -> int:
         return end_line_num
 
     def run(
-            self, tf_file: str, definition_blocks: List[Dict[str, Any]], collect_skip_comments: bool = True
+            self, tf_file: str | TFDefinitionKey, definition_blocks: List[Dict[str, Any]], collect_skip_comments: bool = True
```
Comment (bo156) [inline checkov/terraform/context_parsers/base_parser.py:158]: this one can also be `TFDefinitonKeyType`

File: checkov/terraform/context_parsers/registry.py [hunk 6]
```diff
@@ -26,24 +31,24 @@ def reset_definitions_context(self) -> None:
 
     def enrich_definitions_context(
         self, definitions: Tuple[str, Dict[str, List[Dict[str, Any]]]], collect_skip_comments: bool = True
-    ) -> Dict[str, Dict[str, Dict[str, Any]]]:
+    ) -> Dict[str | TFDefinitionKey, Dict[str, Dict[str, Any]]]:
```
Comment (bo156) [inline checkov/terraform/context_parsers/registry.py:34]: same here

File: checkov/terraform/graph_builder/graph_components/blocks.py [hunk 7]
```diff
@@ -59,7 +59,7 @@ def __init__(
             has_dynamic_block=has_dynamic_block,
             dynamic_attributes=dynamic_attributes,
         )
-        self.module_dependency: str | None = ""
+        self.module_dependency: str | None | TFDefinitionKey = ""
```
Comment (bo156) [inline checkov/terraform/graph_builder/graph_components/blocks.py:62]: same

File: checkov/terraform/context_parsers/registry.py [hunk 8]
```diff
@@ -26,24 +31,24 @@ def reset_definitions_context(self) -> None:
...
         supported_definitions = [parser_type for parser_type in self.context_parsers.keys()]
         (tf_definition_key, definition_blocks_types) = definitions
+        enable_definition_key = strtobool(os.getenv('ENABLE_DEFINITION_KEY', 'False'))
```
Comment (bo156) [inline checkov/terraform/context_parsers/registry.py:37]: I might be missing something, but I'm not familiar with `ENABLE_DEFINITION_KEY` env variable.
The one I know of is `CHECKOV_NEW_TF_PARSER`. where is this one defined?

Comment (lirshindalman) [reply checkov/terraform/context_parsers/registry.py:37]: it's a new one. and its will be modified from the platform under FF

Comment (bo156) [reply checkov/terraform/context_parsers/registry.py:37]: cool :)

**Review by bo156 — APPROVED** [review]
nice work!

File: checkov/common/util/parser_utils.py [hunk 9]
```diff
@@ -375,23 +377,26 @@ def get_tf_definition_key(nested_module: str, module_name: str, module_index: An
 
 
 def get_tf_definition_key_from_module_dependency(
-    path: str, module_dependency: str | None, module_dependency_num: str | None
+    path: str, module_dependency: TFDefinitionKeyType | None, module_dependency_num: str | None
```
Comment (YaaraVerner) [inline checkov/common/util/parser_utils.py:380]: this function should only get called from the old flow (that doesn't use TFDefinitionKeyType), why is this change needed?

Comment (lirshindalman) [reply checkov/common/util/parser_utils.py:380]: You're right. mypy error

File: checkov/terraform/runner.py [hunk 10]
```diff
@@ -569,10 +571,11 @@ def push_skipped_checks_down_old(definition_context, module_path, skipped_checks
     def push_skipped_checks_down_from_modules(self, definition_context):
         module_context_parser = parser_registry.context_parsers[BlockType.MODULE]
         for tf_definition_key, definition in self.definitions.items():
-            if isinstance(tf_definition_key, TFDefinitionKey):
-                full_file_path = tf_definition_key.file_path
+            if not strtobool(os.getenv('ENABLE_DEFINITION_KEY', 'False')):
```
Comment (YaaraVerner) [inline checkov/terraform/runner.py:574]: can you add this to the `EnvVarsConfig`?
